ValidateSelection extension, errors and new estimate fee endpoint test cases

piotr-iohk opened the pull request:
# Issue Number

#469

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have added few cases for estimate fee endpoint (`0` amt tx estimation and estimate in case of `ErrInputDepleted`)
- [x] I have added `validateSelection` to TransationLayer API and corresponding impl
- [x] I have added CLI integration tests illustrating the error for http-bridge
- [x] I have added `ErrValidateSelection` error


# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

Commit:
tag estimate api test to comply with capabilities doc https://docs.google.com/spreadsheets/d/1fadIA_j4nCd3FbylPo5J8K9Fy6tixSC2T9V3xHKuUvU

Commit:
Estimate case for ErrInputsDepleted

Commit:
Zero amount fee test for http-bridge

Commit:
Zero amount fee and post tx tests for jormungandr

piotr-iohk commented on lib/http-bridge/test/integration/Test/Integration/HttpBridge/Scenario/API/Transactions.hs:
```diff
@@ -41,6 +42,34 @@ import qualified Network.HTTP.Types.Status as HTTP
+        (wSrc, payload) <- fixtureZeroAmtSingle ctx
+        r <- request @ApiFee ctx (postTxFeeEp wSrc) Default payload
+        verify r
+            [ expectResponseCode HTTP.status202 ]
```
In case of http-bridge I was rather expecting an error message when trying to estimate fee with `0` amt transaction (the same as it is seen when trying to post transaction with `0` amt -> `expectErrorMessage errMsg403InvalidTransaction`)

@paweljakubas what do you think?

paweljakubas replied:
@piotr-iohk adding error when any output of transaction is 0 upon estimateFee for http-bridge sounds reasonable - it will inform user that this transaction does not have a chance of passing and 0 amt output should be avoided

paweljakubas reviewed (approved):
LGTM! Added `CannotEstimateFee` error that is thrown when 0 amount output is present for http-bridge case only. Also specific CLI integration tests for the aforementioned

paweljakubas committed:
add http bridge specific CLI integration tests for CannotEstimateFee

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -628,12 +635,17 @@ newWalletLayer tracer block0 feePolicy db nw tl = do
         -> NonEmpty TxOut
         -> ExceptT ErrEstimateTxFee IO Fee
-    _estimateTxFee wid opts recipients = do
+    _estimateTxFee backend wid opts recipients = do
```
I am afraid that this isn't an acceptable solution :s ! 
We carry the backend type as the type-level (that's what the `t` is for here, `t` is for `target` and represent the target backend type we are dealing with). We then use class-constraints to grant powers to a given function, making explicit what sort of backend-specific manipulation this function does. 

We can't have that at the value level though, as this will become rapidly unmanageable! In particular, code that is specific to a particular backend _must_ be located and contained into the module of that very backend. If we start putting in the shared piece (a.k.a core) things that are backend-specific this will spread and eventually become out of hand.

If I understand correctly, the issue here is to have some variations in the way null amounts are handled between either backend. Therefore, we need to provide a way for a particular implementation to tell whether a transaction / selection is valid or not. I'd suggest to extend the `TransactionLayer` with a new function: `validateSelection` or something similar which will yield either `()` or a `CreatedInvalidTransaction` (note that there's no need for a new branch `ErrEstimateTxFeeZeroAmountPresent`, which is by the way, maybe a bit too-specific ^^", since we already have an error that entirely captures that very issue). 

This function can then be implemented for both backends which have different rules about transactions. We know for instance that Jormungandr's transactions can't have more than 255 inputs or outputs  and that the bridge can't have output with a null amount. This function can then be called transparently from the wallet layer :), leaving the details out of the wallet core code!

paweljakubas replied:
sound argument. Revisited all stuff!

paweljakubas reviewed (changes requested):
We need to extend TransactionLayer before merging - I will do it in coming commit

paweljakubas committed:
add validateSelection to TransactionLayer

correct validateSelection for jormungandr

correct unit tests

paweljakubas reviewed (approved):
Added validateSelection to TransactionLayer and separate impl for http-bridge (outlawing 0 amount outputs) and  for jormungandr (checking if number of inputs and outputs <= 255)

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -535,6 +535,12 @@ instance LiftHandler ErrEstimateTxFee where
+            apiError err403 CreatedInvalidTransaction $ mconcat
+                [ "I can't estimate transaction fee because either"
+                , " transaction contains at least one payment output of value 0 (Byron)"
+                , " or a transaction has more than 255 inputs or outputs (Jormungandr)."
```
Hmmm. It'd be best not to mention Byron or Jormungandr in the core code. Here, it'd be fine to simply say that the transaction is invalid and give a few examples of invalidity but, with a more general tone. Something like

"It could be that the transaction has too many inputs and/or outputs or, null outputs"

Anviking replied:
Wouldn't a user like to know the exact reason the selection was invalid? Why not have `ErrMkStdTx Text` where `Text` is a user-presentable error with the exact failure-reason?

And if not: wouldn't the user at least want the rules that can make selection invalid? On that bases I think the diff is better than "It could be that the transaction has too many inputs and/or outputs or, null outputs" — on its own. But wrapping the details in a slight general tone might be a good idea.

paweljakubas replied:
done, although separated messages fro `ErrInvalidTxOutAmount` and `ErrExceededInpsOrOuts` of `ErrValidateSelection`

Anviking replied:
From the user-perspective I think this is good.

But `ErrInvalidTxOutAmount` and `ErrExceededInpsOrOuts` are very intertwined with the specific targets.

🤔 I'd be intrigued to let the targets themselves render the error to the user. **E.g**. by using `ErrEstimateTxFeeBackend Text`.

Something like:
<details>

```patch
diff --git a/lib/core/src/Cardano/Wallet.hs b/lib/core/src/Cardano/Wallet.hs
index 4372a1d5..cb4de208 100644
--- a/lib/core/src/Cardano/Wallet.hs
+++ b/lib/core/src/Cardano/Wallet.hs
@@ -32,7 +32,6 @@ module Cardano.Wallet
     , ErrWithRootKey (..)
     , ErrWrongPassphrase (..)
     , ErrMkStdTx (..)
-    , ErrValidateSelection (..)
     , ErrPostTx (..)
     , ErrNetworkUnreachable (..)
     , ErrCoinSelection (..)
@@ -120,7 +119,7 @@ import Cardano.Wallet.Primitive.Types
     , slotRatio
     )
 import Cardano.Wallet.Transaction
-    ( ErrMkStdTx (..), ErrValidateSelection (..), TransactionLayer (..) )
+    ( ErrMkStdTx (..), TransactionLayer (..) )
 import Cardano.Wallet.Unsafe
     ( unsafeRunExceptT )
 import Control.Arrow
@@ -294,7 +293,7 @@ data ErrCreateUnsignedTx
 data ErrEstimateTxFee
     = ErrEstimateTxFeeNoSuchWallet ErrNoSuchWallet
     | ErrEstimateTxFeeCoinSelection ErrCoinSelection
-    | ErrEstimateTxFeeBackend ErrValidateSelection
+    | ErrEstimateTxFeeBackend Text
     deriving (Show, Eq)

 -- | Errors occuring when signing a transaction
diff --git a/lib/core/src/Cardano/Wallet/Api/Server.hs b/lib/core/src/Cardano/Wallet/Api/Server.hs
index 5731a4c8..fdfa667c 100644
--- a/lib/core/src/Cardano/Wallet/Api/Server.hs
+++ b/lib/core/src/Cardano/Wallet/Api/Server.hs
@@ -37,7 +37,6 @@ import Cardano.Wallet
     , ErrSignTx (..)
     , ErrSubmitTx (..)
     , ErrUpdatePassphrase (..)
-    , ErrValidateSelection (..)
     , ErrWalletAlreadyExists (..)
     , ErrWithRootKey (..)
     , ErrWrongPassphrase (..)
@@ -536,15 +535,10 @@ instance LiftHandler ErrEstimateTxFee where
     handler = \case
         ErrEstimateTxFeeNoSuchWallet e -> handler e
         ErrEstimateTxFeeCoinSelection e -> handler e
-        ErrEstimateTxFeeBackend ErrInvalidTxOutAmount ->
+        ErrEstimateTxFeeBackend reason ->
             apiError err403 CannotValidateSelection $ mconcat
-                [ "I can't estimate transaction fee because transaction contains"
-                , " at least one payment output of value 0."
-                ]
-        ErrEstimateTxFeeBackend ErrExceededInpsOrOuts ->
-            apiError err403 CannotValidateSelection $ mconcat
-                [ "I can't estimate transaction fee because transaction has either"
-                , " more than 255 inputs or more than 255 outputs."
+                [ "I can't estimate transaction fee because: "
+                , reason
                 ]

 instance LiftHandler ErrSignTx where
diff --git a/lib/core/src/Cardano/Wallet/Transaction.hs b/lib/core/src/Cardano/Wallet/Transaction.hs
index 92ae1365..f846cbfe 100644
--- a/lib/core/src/Cardano/Wallet/Transaction.hs
+++ b/lib/core/src/Cardano/Wallet/Transaction.hs
@@ -18,7 +18,6 @@ module Cardano.Wallet.Transaction

     -- * Errors
     , ErrMkStdTx (..)
-    , ErrValidateSelection (..)

     -- * Backend helpers
     , estimateMaxNumberOfInputsBase
@@ -35,6 +34,8 @@ import Cardano.Wallet.Primitive.Types
     ( Address (..), Hash (..), Tx, TxIn (..), TxOut (..), TxWitness (..) )
 import Data.Quantity
     ( Quantity (..) )
+import Data.Text
+    ( Text )
 import Data.Word
     ( Word16, Word8 )

@@ -82,7 +83,7 @@ data TransactionLayer t = TransactionLayer

     , validateSelection
       :: CoinSelection
-      -> Either ErrValidateSelection ()
+      -> Either Text ()
       -- ^ Validate coin selection
     }

@@ -95,13 +96,6 @@ data ErrMkStdTx
     -- or when transaction has more than 255 inputs or outputs (Jormungandr)
     deriving (Eq, Show)

-data ErrValidateSelection
-    = ErrInvalidTxOutAmount
-    -- ^ transaction with 0 amount is tried
-    | ErrExceededInpsOrOuts
-    -- ^ transaction has more than 255 inputs or outputs
-    deriving (Eq, Show)
-

 -- | Backend-specific variables used by 'estimateMaxNumberOfInputsBase'.
 data EstimateMaxNumberOfInputsParams t = EstimateMaxNumberOfInputsParams
diff --git a/lib/http-bridge/src/Cardano/Wallet/HttpBridge/Transaction.hs b/lib/http-bridge/src/Cardano/Wallet/HttpBridge/Transaction.hs
index 268fc4f9..fb5d15d6 100644
--- a/lib/http-bridge/src/Cardano/Wallet/HttpBridge/Transaction.hs
+++ b/lib/http-bridge/src/Cardano/Wallet/HttpBridge/Transaction.hs
@@ -40,11 +40,7 @@ import Cardano.Wallet.Primitive.Types
     , txId
     )
 import Cardano.Wallet.Transaction
-    ( ErrMkStdTx (..)
-    , ErrValidateSelection (..)
-    , TransactionLayer (..)
-    , estimateMaxNumberOfInputsBase
-    )
+    ( ErrMkStdTx (..), TransactionLayer (..), estimateMaxNumberOfInputsBase )
 import Control.Monad
     ( forM, when )
 import Data.ByteString
@@ -94,7 +90,7 @@ newTransactionLayer = TransactionLayer

     , validateSelection = \(CoinSelection _ outs _) -> do
         when (any (\ (TxOut _ c) -> c == Coin 0) outs)
-            $ Left ErrInvalidTxOutAmount
+            $ Left "Contains one or more outputs with value 0"
     }
   where
     mkWitness
diff --git a/lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs b/lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs
index d45e8029..a7f4ebc4 100644
--- a/lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs
+++ b/lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs
@@ -24,11 +24,7 @@ import Cardano.Wallet.Primitive.CoinSelection
 import Cardano.Wallet.Primitive.Types
     ( Hash (..), TxOut (..), TxWitness (..), txId )
 import Cardano.Wallet.Transaction
-    ( ErrMkStdTx (..)
-    , ErrValidateSelection (..)
-    , TransactionLayer (..)
-    , estimateMaxNumberOfInputsBase
-    )
+    ( ErrMkStdTx (..), TransactionLayer (..), estimateMaxNumberOfInputsBase )
 import Control.Arrow
     ( second )
 import Control.Monad
@@ -63,8 +59,11 @@ newTransactionLayer (Hash block0) = TransactionLayer
         estimateMaxNumberOfInputsBase @t Binary.estimateMaxNumberOfInputsParams

     , validateSelection = \(CoinSelection inps outs _) -> do
-        when (length inps > 255 || length outs > 255)
-            $ Left ErrExceededInpsOrOuts
+        let validationErr = Left
+        when (length inps > 255) $
+            validationErr "Number of inputs exceeded 255."
+        when (length outs > 255) $
+            validationErr "Number of outputs exceeded 255."
     }
   where
     sign
```

</details>

KtorZ commented on lib/core/src/Cardano/Wallet/Transaction.hs:
```diff
@@ -79,6 +79,10 @@ data TransactionLayer t = TransactionLayer
 
+    , validateSelection
+      :: CoinSelection
+      -> Either ErrMkStdTx ()
```
Can we really return all the error cases from `ErrMkStdTx`? It may be convenient to re-use the same error type but we are probably conveying a misleading information. It'd be better to have another error type `ErrInvalidTx` or something like this.

paweljakubas replied:
Added `ErrValidateSelection` to comply with naming convention introduced by `ErrMkStdTx`

Anviking reviewed (approved):
- Happy to see `validateSelection` in the TransactionLayer.
- Not sure about the way to achieve the target specific errors, but I'd be fine with it too.
- Didn't look much at the integration test / not that familiar with them

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -293,6 +294,7 @@ data ErrCreateUnsignedTx
 data ErrEstimateTxFee
     = ErrEstimateTxFeeNoSuchWallet ErrNoSuchWallet
     | ErrEstimateTxFeeCoinSelection ErrCoinSelection
+    | ErrEstimateTxFeeBackend ErrValidateSelection
```
The term `Backend` is a bit misleading here too. I'd rather stick with the naming convention and have something like:

```hs
ErrEstimateTxFeeValidateSelection ErrValidateSelection
```

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -637,6 +639,8 @@ newWalletLayer tracer block0 feePolicy db nw tl = do
         let utxo = availableUTxO @s @t w
         (sel, _utxo') <- withExceptT ErrEstimateTxFeeCoinSelection $
             CoinSelection.random opts recipients utxo
+        _ <- withExceptT ErrEstimateTxFeeBackend $
```
There's no need for `_ <-` if the function returns `()`. It also makes it obvious that we aren't actually discarding anything.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -208,6 +208,7 @@ data ApiErrorCode
     | CannotCoverFee
     | NetworkUnreachable
     | CreatedInvalidTransaction
+    | CannotValidateSelection
```
Hmmm. The formulation is a bit misleading here. It's not that we can't validate the selection, it's really that the selection is not valid.  Also, this error code seems redundant with `CreatedInvalidTransaction` (cf other review comment about transaction layer consistency)

paweljakubas replied:
Agree, `CreatedInvalidTransaction` should be reused here

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -535,6 +536,16 @@ instance LiftHandler ErrEstimateTxFee where
+        ErrEstimateTxFeeBackend ErrExceededInpsOrOuts ->
+            apiError err403 CannotValidateSelection $ mconcat
+                [ "I can't estimate transaction fee because transaction has either"
+                , " more than 255 inputs or more than 255 outputs."
```
It'd be better to have these numbers (255) encoded within the error type since this is actually configuration / backend-specific. Some like: 

```hs
ErrExceededInpsOrOuts = ErrExceededInpsOrOuts
    { maxInps :: Int
    , maxOuts :: Int
    }
```

KtorZ commented on lib/core/test/integration/Test/Integration/Framework/TestData.hs:
```diff
@@ -245,6 +246,10 @@ errMsg403InvalidTransaction :: String
 
+errMsg403CannotEstimateFee :: String
+errMsg403CannotEstimateFee = "I can't estimate transaction fee because transaction contains\
+    \ at least one payment output of value 0."
```
We're probably going to need two different messages here right, one for each case?

KtorZ replied:
Not sure how we can manage to test the other case though ....

paweljakubas replied:
two messages here, one is already caught in http-bridge specific integration tests, the other will be caught on specific jormunagandr integration tests (and when the corresponding case will be added as a second value)

paweljakubas replied:
I unified `errMsg403InvalidTransaction` with `errMsg403CannotEstimateFee` as not coin selection validation is used (usisng specific implementation for both backends)

KtorZ commented on lib/http-bridge/src/Cardano/Wallet/HttpBridge/Transaction.hs:
```diff
@@ -88,6 +92,9 @@ newTransactionLayer = TransactionLayer
 
+    , validateSelection = \(CoinSelection _ outs _) -> do
+        when (any (\ (TxOut _ c) -> c == Coin 0) outs)
+            $ Left ErrInvalidTxOutAmount
```
There's now a logic duplication regarding tx validation. The same thing is done in `mkStdTx` and that's something we don't want to be duplicated for there's a risk of diverging as we maintain the code.
My suggestion:

- Change `ErrInvalidTx` to:

```hs
-- | Possible signing error
data ErrMkStdTx
    = ErrKeyNotFoundForAddress Address
    -- ^ We tried to sign a transaction with inputs that are unknown to us?
    | ErrInvalidTx ErrValidateSelection
    -- ^ when transaction with 0 amount is tried (not valid in Byron)
    deriving (Eq, Show)
```

- Call `validateSelection` from within `mkStdTx` and wrap the error accordingly.

paweljakubas replied:
yes, it is very sensible to do it

KtorZ commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs:
```diff
@@ -56,6 +62,9 @@ newTransactionLayer (Hash block0) = TransactionLayer
 
+    , validateSelection = \(CoinSelection inps outs _) -> do
+        when (length inps > 255 || length outs > 255)
+            $ Left ErrExceededInpsOrOuts
```
Same remark about code duplication applies here too. Plus, this check doesn't even occur at the moment in `mkStdTx`. 
Also, since we are starting to reference this `255` in various places in Jörmungandr, it'd be good to have them as named constant somewhere (probably in the `Binary` module as `maxNumberOfInputs` and `maxNumberOfOutputs`).

piotr-iohk replied:
Yes, TX with 256 inputs now ends up with error `500` and `Something went wrong` message. Should be addressed similarly to the fee estimation.

paweljakubas replied:
added in jormungandr' Binary module both `maxNumberOfInputs` and `maxNumberOfOutputs`

KtorZ commented on lib/core/test/integration/Test/Integration/Framework/TestData.hs:
```diff
@@ -241,9 +243,14 @@ errMsg403InputsDepleted = "I had to select inputs to construct the requested\
-errMsg403InvalidTransaction = "I can't process this payment because it contains at least\
-    \ one payment output of value 0. This isn't supported by the current core nodes."
+errMsg403InvalidTransactionZeroAmtOutput :: String
+errMsg403InvalidTransactionZeroAmtOutput = "I can't validate coin selection because transaction contains\
```
The message sounds a bit weird, we first talk about `coin selection` and then `transaction`. They are indeed closed concepts, but using them interchangeably feels a bit confusing. What about replacing transaction by "_it_" or something that shows better the relationship between tx and selection:

```
I can't construct a valid transaction from the given coin selection because it contains at least one payment output of value 0.
```

paweljakubas replied:
Change it in such a way that transaction is not mentioned. This is because we use the same error both when constructing transaction and when estimating fee

KtorZ commented on lib/core/test/integration/Test/Integration/Framework/TestData.hs:
```diff
@@ -241,9 +243,14 @@ errMsg403InputsDepleted = "I had to select inputs to construct the requested\
+
+_errMsg403InvalidTransactionWrongInpsOuts :: MaxInpsOrOuts -> String
+_errMsg403InvalidTransactionWrongInpsOuts (MaxInpsOrOuts maxNumInps maxNumOuts) =
+    "I can't validate coin selection because transaction has either\
```
Same remark about the relationship between coin selection and transaction in the message.

paweljakubas replied:
Change it in such a way that transaction is not mentioned. This is because we use the same error both when constructing transaction and when estimating fee

KtorZ commented on lib/core/test/unit/Cardano/Wallet/TransactionSpec.hs:
```diff
@@ -15,7 +15,6 @@ spec :: Spec
 spec =
     describe "Pointless tests to cover 'Show' instances for errors" $ do
         testShow $ ErrKeyNotFoundForAddress $ Address mempty
-        testShow ErrInvalidTx
```
Why is this gone :thinking: ?

paweljakubas replied:
`ErrInvalidTx` got more complicated and in description I see **pointless tests...** ...now added it back

KtorZ commented on lib/core/test/unit/Cardano/WalletSpec.hs:
```diff
@@ -366,6 +366,7 @@ dummyTransactionLayer = TransactionLayer
         error "dummyTransactionLayer: estimateSize not implemented"
     , estimateMaxNumberOfInputs =
         error "dummyTransactionLayer: estimateMaxNumberOfInputs not implemented"
+    , validateSelection = \_ -> pure ()
```
It'd be better to actually throw an `error not implemented` here too. Providing a (wrong) default implementation may lead to unpleasant surprise if we suddenly start using this branch. If we haven't implemented this, let's say it clear and loud :)

paweljakubas replied:
previously as validateSelection was used in `_signTx` the above hack was needed. now as we moved everything to `mkStdTx` we can provide `error not implemented`

KtorZ commented on lib/http-bridge/src/Cardano/Wallet/HttpBridge/Transaction.hs:
```diff
@@ -65,8 +69,6 @@ newTransactionLayer = TransactionLayer
         let tx = Tx ins outs
-        when (any (\ (TxOut _ c) -> c == Coin 0) outs)
-            $ Left ErrInvalidTx
         let txSigData = txId @(HttpBridge n) tx
```
So, `validateSelection` is now called from the upper layer in `Cardano.Wallet`, though it should really be called from within `mkStdTx`. Otherwise, it becomes the caller responsibility to know that they have to validate the selection before calling `mkStdTx` which may cause hard to catch errors later on. Since we don't guarantee at the type-level that a selection is valid, then, we have to put this check / invariant in the `mkStdTx` function where it actually matters for the selection to be valid!

paweljakubas replied:
now `validateSelection` used inside `mkStdTx`

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -646,11 +655,13 @@ newWalletLayer tracer block0 feePolicy db nw tl = do
                 addr <- state genChange
                 return $ TxOut addr c
+        withExceptT (ErrSignTx . ErrInvalidTx) $
+            except $ (validateSelection tl) sel
```
I think this is now in the wrong place (should be part of `mkStdTx` function), will confirm while reviewing the rest.

KtorZ commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs:
```diff
@@ -56,6 +66,10 @@ newTransactionLayer (Hash block0) = TransactionLayer
+    , validateSelection = \(CoinSelection inps outs _) -> do
+        when (length inps > maxNumberOfInputs || length outs > maxNumberOfOutputs)
+            $ Left $ ErrExceededInpsOrOuts
+            (MaxInpsOrOuts maxNumberOfInputs maxNumberOfOutputs)
```
:+1:

KtorZ commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:
```diff
@@ -133,6 +134,15 @@ import qualified Data.ByteArray as BA
+
+-- maximum number of outputs in a valid transaction
+maxNumberOfOutputs :: Int
+maxNumberOfOutputs = 255
```
:+1:
I think however that there are a few places in this very module which also use these constants in a hard-coded form. It'd be good if we could replace them too by using their named equivalent newly created :smile:

paweljakubas replied:
found two more places where 255 is used

KtorZ committed:
Move coin selection validation down to the coin selection functions.

After thoughts, it is a bit weird to have the coin selection yield an
invalid coin selection, and later on only check whether this selection
is valid. Instead, we can pretty much ensure that the output produced by
the coin selection is valid IN THE FIRST PLACE.

Still, since the validation is actually backend specific, it requires a
bit of juggling with the type-system. I went for a type family and a
custom error type based on the `t` parameter since this is probably the
easiest and most flexible thing here.

KtorZ committed:
Review 'validateSelection' impl for both backend target and relocate errors there

KtorZ committed:
update core unit tests and add tests about extra CS validation guard
